### ISSUE TITLE
docker-machine failed with error when creating centOS based VM on Azure

### ISSUE THREAD

**Issue by Lickkylee** [docker/machine#3741]
I met errors when using docker-machine to create a centos docker VM on Azure. Here was what I did:

1. I tried centOS 7.2 and the error is:
$ docker-machine.exe create --driver azure --azure-subscription-id "...." --azure-image "OpenLogic:CentOS:7.2:latest" --azure-location "China North" --azure-resource-group "lqidocktestrg" --azure-size "Standard_A2" --azure-vnet "lqidockvnet" --azure-subnet "lqidocksub" --azure-subnet-prefix "192.168.0.0/24" --azure-availability-set "lqidocava" --azure-open-port "80" --azure-private-ip-address  192.168.0.5 --azure-environment "AzureChinaCloud" lqidocvm3
Running pre-create checks...
(lqidocvm3) Registering subscription to resource provider.  ns="Microsoft.Network" subs="..."
(lqidocvm3) Completed machine pre-create checks.
Creating machine...
(lqidocvm3) Querying existing resource group.  name="lqidocktestrg"
(lqidocvm3) Resource group "lqidocktestrg" already exists.
(lqidocvm3) Configuring availability set.  name="lqidocava"
(lqidocvm3) Configuring network security group.  name="lqidocvm3-firewall" location="China North"
(lqidocvm3) Querying if virtual network already exists.  name="lqidockvnet" location="China North"
(lqidocvm3) Virtual network already exists.  name="lqidockvnet" location="China North"
(lqidocvm3) Configuring subnet.  vnet="lqidockvnet" cidr="192.168.0.0/24" name="lqidocksub"
(lqidocvm3) Creating public IP address.  name="lqidocvm3-ip" static=false
(lqidocvm3) Creating network interface.  name="lqidocvm3-nic"
(lqidocvm3) Creating storage account.  name="vhdsgsj2tmb43tp9jpw04us0" location="China North"
(lqidocvm3) Creating virtual machine.  name="lqidocvm3" location="China North" size="Standard_A2" username="docker-user" osImage="OpenLogic:CentOS:7.2:latest"
Waiting for machine to be running, this may take a few minutes...
Detecting operating system of created instance...
Waiting for SSH to be available...
Detecting the provisioner...
Provisioning with centos...
Copying certs to the local machine directory...
Copying certs to the remote machine...
Setting Docker configuration on the remote daemon...
Error creating machine: Error running provisioning: Something went wrong running an SSH command!
command : sudo systemctl -f start docker
err     : exit status 6
output  : Failed to start docker.service: Unit docker.socket failed to load: No such file or directory.

2. I tried centos 6.5 and the error is 
$ docker-machine.exe create --driver azure --azure-subscription-id "9ef8a15c-15a2-4ef1-a19b-e31876ab177c" --azure-image "OpenLogic:CentOS:6.5:latest" --azure-location "China North" --azure-resource-group "lqidocktestrg" --azure-size "Standard_A2" --azure-vnet "lqidockvnet" --azure-subnet "lqidocksub" --azure-subnet-prefix "192.168.0.0/24" --azure-availability-set "lqidocava" --azure-open-port "80" --azure-private-ip-address  192.168.0.22 --azure-environment "AzureChinaCloud" lqidocvm6
Running pre-create checks...
(lqidocvm6) Registering subscription to resource provider.  ns="Microsoft.Network" subs="..."
(lqidocvm6) Completed machine pre-create checks.
Creating machine...
(lqidocvm6) Querying existing resource group.  name="lqidocktestrg"
(lqidocvm6) Resource group "lqidocktestrg" already exists.
(lqidocvm6) Configuring availability set.  name="lqidocava"
(lqidocvm6) Configuring network security group.  name="lqidocvm6-firewall" location="China North"
(lqidocvm6) Querying if virtual network already exists.  name="lqidockvnet" location="China North"
(lqidocvm6) Virtual network already exists.  name="lqidockvnet" location="China North"
(lqidocvm6) Configuring subnet.  name="lqidocksub" vnet="lqidockvnet" cidr="192.168.0.0/24"
(lqidocvm6) Creating public IP address.  name="lqidocvm6-ip" static=false
(lqidocvm6) Creating network interface.  name="lqidocvm6-nic"
(lqidocvm6) Creating storage account.  name="vhdsntzaxwdx2smuy41jkfuk" location="China North"
(lqidocvm6) Creating virtual machine.  size="Standard_A2" username="docker-user" osImage="OpenLogic:CentOS:6.5:latest" name="lqidocvm6" location="China North"
Waiting for machine to be running, this may take a few minutes...
Detecting operating system of created instance...
Waiting for SSH to be available...
Detecting the provisioner...
Error creating machine: Error detecting OS: Error getting SSH command: Something went wrong running an SSH command!
command : cat /etc/os-release
err     : exit status 1
output  : cat: /etc/os-release: No such file or directory

3. However, When i used the same command with Ubuntu, it was okay. 
$ docker-machine.exe create --driver azure --azure-subscription-id "..." --azure-image "Canonical:UbuntuServer:16.04.0-LTS:latest" --azure-location "China North" --azure-resource-group "lqidocktestrg" --azure-size "Standard_A2" --azure-vnet "lqidockvnet" --azure-subnet "lqidocksub" --azure-subnet-prefix "192.168.0.0/24" --azure-availability-set "lqidocava" --azure-open-port "80" --azure-private-ip-address  192.168.0.33 --azure-environment "AzureChinaCloud" lqidocvm7
Running pre-create checks...
(lqidocvm7) Registering subscription to resource provider.  ns="Microsoft.Network" subs="..."
(lqidocvm7) Completed machine pre-create checks.
Creating machine...
(lqidocvm7) Querying existing resource group.  name="lqidocktestrg"
(lqidocvm7) Resource group "lqidocktestrg" already exists.
(lqidocvm7) Configuring availability set.  name="lqidocava"
(lqidocvm7) Configuring network security group.  name="lqidocvm7-firewall" location="China North"
(lqidocvm7) Querying if virtual network already exists.  name="lqidockvnet" location="China North"
(lqidocvm7) Virtual network already exists.  name="lqidockvnet" location="China North"
(lqidocvm7) Configuring subnet.  vnet="lqidockvnet" cidr="192.168.0.0/24" name="lqidocksub"
(lqidocvm7) Creating public IP address.  name="lqidocvm7-ip" static=false
(lqidocvm7) Creating network interface.  name="lqidocvm7-nic"
(lqidocvm7) Creating storage account.  name="vhdsjoogwhhz0lt5fon3oyjn" location="China North"
(lqidocvm7) Creating virtual machine.  name="lqidocvm7" location="China North" size="Standard_A2" username="docker-user" osImage="Canonical:UbuntuServer:16.04.0-LTS:latest"
Waiting for machine to be running, this may take a few minutes...
Detecting operating system of created instance...
Waiting for SSH to be available...
Detecting the provisioner...
Provisioning with ubuntu(systemd)...
Installing Docker...
Copying certs to the local machine directory...
Copying certs to the remote machine...
Setting Docker configuration on the remote daemon...
Checking connection to Docker...
Docker is up and running!
To see how to connect your Docker Client to the Docker Engine running on this virtual machine, run: C:\Users\lqi.FAREAST\bin\docker-machine.exe env lqidocvm7

So my question is: does docker-machine support CentOS ? or it just has limited OS version support of Linux? Where I can find the support information?

**Comment by mbentley** [comment]
I have the same issue:

```
docker-machine create -d azure --azure-size "Standard_A2" --azure-location "centralus" --azure-image OpenLogic:CentOS:7.2:latest --azure-resource-group mbentley-machinetest --azure-availability-set mbentley-machinetest-as --azure-vnet mbentley-machinetest-vnet --azure-ssh-user mbentley --azure-subscription-id <mysub> mbentley-machinetest
Running pre-create checks...
(mbentley-machinetest) Completed machine pre-create checks.
Creating machine...
(mbentley-machinetest) Querying existing resource group.  name="mbentley-machinetest"
(mbentley-machinetest) Creating resource group.  name="mbentley-machinetest" location="centralus"
(mbentley-machinetest) Configuring availability set.  name="mbentley-machinetest-as"
(mbentley-machinetest) Configuring network security group.  location="centralus" name="mbentley-machinetest-firewall"
(mbentley-machinetest) Querying if virtual network already exists.  name="mbentley-machinetest-vnet" location="centralus"
(mbentley-machinetest) Configuring subnet.  name="docker-machine" vnet="mbentley-machinetest-vnet" cidr="192.168.0.0/16"
(mbentley-machinetest) Creating public IP address.  name="mbentley-machinetest-ip" static=false
(mbentley-machinetest) Creating network interface.  name="mbentley-machinetest-nic"
(mbentley-machinetest) Creating storage account.  name="vhdsmb4k9dkrnj9exlp386ek" location="centralus"
(mbentley-machinetest) Creating virtual machine.  name="mbentley-machinetest" location="centralus" size="Standard_A2" username="mbentley" osImage="OpenLogic:CentOS:7.2:latest"
Waiting for machine to be running, this may take a few minutes...
Detecting operating system of created instance...
Waiting for SSH to be available...
Detecting the provisioner...
Provisioning with centos...
Copying certs to the local machine directory...
Copying certs to the remote machine...
Setting Docker configuration on the remote daemon...
Error creating machine: Error running provisioning: Something went wrong running an SSH command!
command : sudo systemctl -f start docker
err     : exit status 6
output  : Failed to start docker.service: Unit docker.socket failed to load: No such file or directory.
```

For whatever reason, `docker.socket` is most definitely missing from the host.

**Comment by ahmetalpbalkan** [comment]
@Lickkylee @mbentley First of all I have good reason to believe this is not related to Azure (as the error happens after `Detecting the provisioner...`).

`get.docker.com` Docker installation script (which also happens to be what `docker-machine` uses to install docker-engine) recently removed socket-activation from CentOS part of the script: https://github.com/docker/docker/issues/23793

I am guessing the `centos` provisioner in docker-machine is not updated to reflect these changes and I'm also guessing you'd get this error on any other cloud provider with the centos7.2 images. 

Basically the `get.docker.com` script has broken the downstream yet again (in this case the docker-machine). Personally I've worked around this issue in another project I manage (azure-docker-extension) and that's how I particularly know about this problem.

**Comment by Lickkylee** [comment]
@ahmetalpbalkan Thanks for your comments. Yes, you are right, it's not a issue on Azure. I believed this was a bug of docker-machine since when I used latest version 0.8.2 I could succeed in creating CentOS VM and starting docker service using Azure Driver. Will change the title.

**Comment by mbentley** [comment]
Did you mean to close this issue?
